extmod/asyncio: Add ssl support with SSLContext.

Carglglz opened the pull request:
This adds `asyncio` ssl support ~~and fix async stream read methods in unix port.
The reason is stream ssl sockets seem to behave different from plain stream sockets. The main difference
is doing partial reads on ssl sockets will clear any write/read/close "event" flag even if there is still
pending data to be read. So registering the ssl socket again in the poller will wait there until a new event
or timeout.
The fix is check first if the ssl socket has pending data and read/return it instead of registering in the poller to wait
for an event.~~

Follow-up to:
- [x] #13098
~- [ ] #11896~ 

This may close #8647, #9071, #8177, #5840, #7315, #9015

Carglglz commented:
Note that the "fix" is more a workaround than a real fix, I'm not sure where in the codebase the "read/write/close" events for stream ssl sockets are handled other than in `modssl_mbedtls.c` and why the bug affects the unix port but not the esp32 port...

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -30,6 +36,15 @@ async def wait_closed(self):
     def read(self, n=-1):
         r = b""
         while True:
+            if hasattr(self.s, "ssl_pending"):
```
Is this `ssl_pending` only needed for the unix port to work correctly?  If so I would rather fix unix by making it work correctly when polling an ssl socket.

dpgeorge replied:
I can confirm that the `tests/multi_net/asyncio_tls_server_client_readline.py` test works with the `ssl_pending` code there and fails when it is removed.

But, by enabling `MICROPY_PY_SELECT`, it will work without the `ssl_pending` code.

So I will fix the unix port so that it can work without needing to enable that option.

BTW, to run this test I actually had to regenerate the certificate and key, because otherwise it gave a "certificate has expired" error.

Carglglz replied:
> But, by enabling MICROPY_PY_SELECT, it will work without the ssl_pending code.

well that seems like an easy fix 😅

> BTW, to run this test I actually had to regenerate the certificate and key, because otherwise it gave a "certificate has expired" error.

I will update the tests too with a longer expiration date then 👍🏼

Carglglz replied:
> But, by enabling MICROPY_PY_SELECT, it will work without the ssl_pending code.
So I will fix the unix port so that it can work without needing to enable that option

Will wait for the fix then... 👀

dpgeorge replied:
#12138 is now merged, which should fix the unix port so it doesn't need special handling for ssl-wrapped sockets.

dpgeorge replied:
I found the issue with the readline test failing here, and #12195 should fix that.

dpgeorge replied:
The above fix is now in master, so if you could please rebase this PR and see if the CI passes.

Also can you please re-test asyncio-mqtt.simple and asyncio-microdot.  If they don't work then please provide details how to reproduce the error.  Thanks!

Carglglz replied:
> Also can you please re-test asyncio-mqtt.simple and asyncio-microdot. If they don't work then please provide details how to reproduce the error. Thanks!

It's working again now 😮‍💨👍🏼 ! It must be #12195, so really nice work figuring this out  👏🏼


> The above fix is now in master, so if you could please rebase this PR and see if the CI passes

Done! and CI tests are passing now too.

dpgeorge replied:
Very good, thanks for testing!

Carglglz replied:
After some further testing I've found another weird bug in unix port 👀, this time is the combination of a stream ssl socket and a plain stream socket .i.e running two `asyncio-microdot` web servers instances,  one with HTTP and the other with HTTPS. What happens is that at first both work BUT the moment I do a request to the HTTP instance,  then it will block the requests to the HTTPS instance until I do another request to the HTTP instance. 

I've also tested this with .e.g TASK A -> `async_urequests.py` (HTTP request) and TASK B ->`asyncio-mqtt.simple` (SSL) with the same results, i.e. `mqtt` client stops reading (writing still works) after the TASK A does an HTTP request, and then it will only work after TASK A does another HTTP request and so on.

The combination of HTTP-HTTP and HTTPS-HTTPS is OK. 
esp32 port is OK
Probably there is another case/flag missing like in #12195 which I'm not aware of 🤷🏼‍♂️

Carglglz replied:
I've updated `tests/net_inet` with `asyncio_http_https_opc.py` and `asyncio_https_http_opc.py` that reflects this behaviour, .i.e first test gets stuck after http request, second test works as expected. 
```
$ micropython asyncio_https_http_opc.py
write GET
read response
read: b'HTTP/1.1 200 OK'
close
done
write GET
read response
read: b'HTTP/1.1 200 OK'
close
done
$ micropython asyncio_http_https_opc.py
write GET
read response
read: b'HTTP/1.1 200 OK'
close
done
^CTraceback (most recent call last):
  File "asyncio_http_https_opc.py", line 109, in <module>
  File "asyncio/core.py", line 1, in run
  File "asyncio/core.py", line 1, in run_until_complete
  File "asyncio/core.py", line 1, in wait_io_event
KeyboardInterrupt:

```

Carglglz replied:
I've added `tests/multi_net/asyncio_tcp_tls_server_client.py` and `tests/multi_net/asyncio_tls_tcp_server_client.py` 

These tests cover the above edge case and as you can see `tests/multi_net/asyncio_tcp_tls_server_client.py` is failing...

Carglglz replied:
I think I've found what's going on with the HTTP request, it turns out that the function `poll_set_poll_once` https://github.com/micropython/micropython/blob/24a6e951ec7696b8d18d95fe5da36f7e489913d0/extmod/modselect.c#L252 in `extmod/modselect.c` gets called in an infinite loop after the first HTTP request. Doing an HTTPS request stops that loop but it also gets stuck until another HTTP request which starts the infinite loop again. e.g.

```
# HTTPS REQUEST
POLL
POLL
Request line
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR
POLL
POLL
POLL
POLL
IOCTL REQUEST: 3
IOCTL REQUEST TO SOCKET
IOCTL REQUEST SOCK: 3
GET /temp HTTP/1.1
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR
POLL
POLL
POLL
POLL
IOCTL REQUEST: 3
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR
POLL
POLL
POLL
POLL
IOCTL REQUEST: 3
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR
POLL
POLL
POLL
POLL
IOCTL REQUEST: 3
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR
POLL
POLL
POLL
POLL
IOCTL REQUEST: 3
IOCTL REQUEST: 4
IOCTL REQUEST TO SOCKET
IOCTL REQUEST SOCK: 4
2023-08-12 17:43:35 [darwin@macos] [INFO] [microdot.service] GET /temp 200
# HTTP REQUEST
POLL
POLL
POLL
POLL
POLL
POLL
POLL
POLL
POLL
2023-08-12 17:46:32 [darwin@macos] [INFO] [stats.service] GET /debug HTTP/1.1
2023-08-12 17:46:32 [darwin@macos] [INFO] [stats.service] HTTP/1.1 200 OK
POLL
....
POLL # (NON STOP UNTIL NEX HTTPS REQUEST)
# HTTPS REQUEST
Request line
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR # (STUCK HERE UNTIL NEXT HTTP REQUEST)

```
So it's definitely related to #12138 and I have no idea how to fix this but I hope these hints can help 👍🏼

Carglglz replied:
I think I've found what's wrong and my conclusion is that `poll_set_all_are_fds(poll_set)` in `extmod/modselect.c` does not work as expected. I've change that for this function 
```C
+// For each object in the poll set check if it has a file descriptor
+//
+#if MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+STATIC bool poll_set_check_all_fd(poll_set_t *poll_set, size_t *rwx_num) {
+
+    for (mp_uint_t i = 0; i < poll_set->map.alloc; ++i) {
+        if (!mp_map_slot_is_filled(&poll_set->map, i)) {
+            continue;
+        }
+
+        poll_obj_t *poll_obj = MP_OBJ_TO_PTR(poll_set->map.table[i].value);
+
+
+        if (poll_obj->pollfd == NULL) {
+            // Object does not have a file descriptor so will be polled separately by poll_set_poll_once().
+            return false;
+        }
+    }
+    return true;
+}
+#endif
+
```
Which checks if all poll objects in the poll set have a file descriptor or not. This fix the mentioned bug above 👍🏼 
I could do a separate PR for this if necessary. 👀

dpgeorge replied:
Yes I see the problem.  Good work finding it!

But I think it should be solved a different way.  I'll push a fix with a test.

Carglglz replied:
FWIW I did try to solve it fixing the `poll_set_all_are_fds` logic checking the `poll_set` length but that turned out to be a slippery slope... 😕 so this was the only reliable method I could think of instead...

dpgeorge replied:
See #12224 for a fix.

dpgeorge replied:
That PR is now merged to master.

Carglglz replied:
Rebased and all CI tests passing, I guess this is finally ready!

Carglglz commented:
OK I rebased this set of PRs on #8988 and it seems like the problem is solved at `mbedtls v3.3.0`, i.e no need to enable `MICROPY_PY_SELECT` (which also looks like it doesn't work here...)

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -63,7 +69,10 @@ def readline(self):
+            if l2:
+                l += l2
+            if l2 is None:
+                continue
```
I think this can be simplified to just:
```py
l2 = self.s.readline()
if l2 is None:
    continue
l += l2
...
```

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -100,20 +109,40 @@ def drain(self):
+        s.setblocking(False)
+        ss = Stream(s)
     try:
         s.connect(ai[-1])
```
For ssl connections, this will be a blocking connect, which is bad.  The stream should be set to non-blocking for both ssl and non-ssl connections.

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -100,20 +109,40 @@ def drain(self):
+            try:
+                assert isinstance(ssl, _ssl.SSLContext)
+            except:
+                raise ValueError("Invalid ssl param")
```
I don't think it's worth having this check (which costs code size).  Instead, just remove this entire `else` block, and if the user doesn't pass in an SSLContext object then they will get a reasonable error when trying to access the `wrap_socket` method below.

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -100,20 +109,40 @@ def drain(self):
+                raise ValueError("Invalid ssl param")
+        if not server_hostname:
+            server_hostname = host
+        s = ssl.wrap_socket(s, server_hostname=server_hostname)
```
I think it should pass through `do_handshake_on_connect=False` to have async handshakes.

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -146,14 +175,24 @@ async def _serve(self, s, cb):
                 continue
+            if s2:
+                if ssl and _ssl:
+                    if isinstance(ssl, _ssl.SSLContext):
```
I don't think it's necessary to have all these checks.  Just do simply `if ssl:`.

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -146,14 +175,24 @@ async def _serve(self, s, cb):
+                    if isinstance(ssl, _ssl.SSLContext):
+                        s2.setblocking(True)
+                        try:
+                            s2 = ssl.wrap_socket(s2, server_side=True)
```
Can it use `do_handshake_on_connect=False`?  And then make the socket non-blocking.

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -146,14 +175,24 @@ async def _serve(self, s, cb):
+                            s2 = ssl.wrap_socket(s2, server_side=True)
+                        except OSError as e:
+                            sys.print_exception(e)
+                            s2.close()
```
Does this behaviour match CPython, of printing the exception to stdout and then continuing?

Carglglz replied:
Not sure about how CPython manages this but after some testing this was the less resource intensive method I could think of, e.g.  the other alternative 
```python
while True:
    try:
          await asyncio.start_server(...)
     except Exception:
            ....
```
creates a `Server` instance and a task every time a handshake fails due to the client using HTTP  or not trusting the server certificate....

Carglglz commented:
I've just tested making the socket non-blocking with `do_handshake_on_connect=False` (on unix port) and it seems to work with HTTPS in both client and server modes. (I think this wouldn't work without the recent fixes or at least that was what I remember)
However the `async_mqtt.py` implementation gets stuck after connection in the first `read` statement... I will need to investigate this further...

Carglglz commented:
I think I fixed it, as I say in the commit message:
> Not doing the handshake on connect will delegate the handshake to the
following `mbedtls_ssl_read/write` calls. However if the handshake fails
the peer needs to be notified by `ssl_socket_read/write` and then raise
the proper mbedtls error code and message.

The drop in coverage comes from the multiple cases in which a handshake may fail, I've already added a test for the CN mismatch case so I think it shouldn't be necessary to cover the rest (they are just value comparisons) .i.e.
at https://github.com/micropython/micropython/blob/e193cdb991e6fd0c3d67088086df364f2601b397/extmod/modssl_mbedtls.c#L160
```C
if (flags & MBEDTLS_X509_BADCERT_OTHER) {
alert = MBEDTLS_SSL_ALERT_MSG_ACCESS_DENIED;
} else if (flags & MBEDTLS_X509_BADCERT_CN_MISMATCH) {
alert = MBEDTLS_SSL_ALERT_MSG_BAD_CERT;
} else if (flags & MBEDTLS_X509_BADCERT_KEY_USAGE) {
alert = MBEDTLS_SSL_ALERT_MSG_UNSUPPORTED_CERT;
} else if (flags & MBEDTLS_X509_BADCERT_EXT_KEY_USAGE) {
alert = MBEDTLS_SSL_ALERT_MSG_UNSUPPORTED_CERT;
} else if (flags & MBEDTLS_X509_BADCERT_NS_CERT_TYPE) {
alert = MBEDTLS_SSL_ALERT_MSG_UNSUPPORTED_CERT;
} else if (flags & MBEDTLS_X509_BADCERT_BAD_PK) {
alert = MBEDTLS_SSL_ALERT_MSG_UNSUPPORTED_CERT;
} else if (flags & MBEDTLS_X509_BADCERT_BAD_KEY) {
alert = MBEDTLS_SSL_ALERT_MSG_UNSUPPORTED_CERT;
} else if (flags & MBEDTLS_X509_BADCERT_EXPIRED) {
alert = MBEDTLS_SSL_ALERT_MSG_CERT_EXPIRED;
} else if (flags & MBEDTLS_X509_BADCERT_REVOKED) {
alert = MBEDTLS_SSL_ALERT_MSG_CERT_REVOKED;
} else if (flags & MBEDTLS_X509_BADCERT_NOT_TRUSTED) {
alert = MBEDTLS_SSL_ALERT_MSG_UNKNOWN_CA;
} else {
alert = MBEDTLS_SSL_ALERT_MSG_CERT_UNKNOWN;
}
} else if (*errcode == MBEDTLS_ERR_SSL_NO_CLIENT_CERTIFICATE) {
  
```
I'm not 100% sure if the  `ssl_handle_handshake_error` bit is at the best place so check it out and let me know what you think👍🏼

dpgeorge commented:
Thanks for investigating this.  I took a look at your latest commit that calls `mbedtls_ssl_send_alert_message()`, and the test you added there, `multi_net/asyncio_tls_server_client_verify_error.py`.

That test passes on the unix port.  But if I remove the code you added to `modssl_mbedtls.c` in that last commit, that test still passes (although the error messages are now integers, not nice strings).

So I'm wondering why you need to call `mbedtls_ssl_send_alert_message()` at all?

> However the async_mqtt.py implementation gets stuck after connection in the first read statement...

I suspect that's the reason why `mbedtls_ssl_send_alert_message()` is needed.  Can you please tell me exactly how to run that mqtt test so I can reproduce this problem?

Carglglz commented:
> So I'm wondering why you need to call mbedtls_ssl_send_alert_message() at all?

I added that to match the behaviour of `mbedtls_ssl_handshake` that happens when `do_handshake_on_connect` is true. Inside `mbedtls_ssl_handshake` if the handshake fails, it notifies the peer with the failure reason. 

I've realised this while testing `microdot + asyncio + tls`, and on client side using `curl` ,`ab`, and CPython `requests`. I tested the handshake failure and the client side got stuck until timeout. 
So then I added the `mbedtls_ssl_send_alert_message` to notify the client side with the proper alert message and worked as expected. 

> >However the async_mqtt.py implementation gets stuck after connection in the first read statement...

>I suspect that's the reason why mbedtls_ssl_send_alert_message() is needed. Can you please tell me exactly how to run that mqtt test so I can reproduce this problem?

That I think, was actually related to not making the handshake right away and the socket being not blocking. I "fixed" it with a workaround. 
```diff
+        self.a_writer.s.setblocking(True)
         self.a_writer.s.write(premsg, i + 2) 
+        self.a_writer.s.setblocking(False)

```
 Note that I ported the mqtt client from micropython-lib and AFAIK there is no asyncio equivalent to `socket.write(premsg, i + 2)` which I also think is not documented? (the second argument)

The MQTT asyncio client is here https://github.com/Carglglz/micropython-lib/blob/async-mqtt-simple/micropython/umqtt.simple/umqtt/async_simple.py

> That test passes on the unix port. But if I remove the code you added to modssl_mbedtls.c in that last commit, that test still passes (although the error messages are now integers, not nice strings).

This is to match the blocking handshake failure too 👍🏼

dpgeorge commented:
> I've realised this while testing `microdot + asyncio + tls`, and on client side using `curl` ,`ab`, and CPython `requests`. I tested the handshake failure and the client side got stuck until timeout.

Are you able to provide some code that shows how to reproduce this particular failure?

> self.a_writer.s.write(premsg, i + 2) 

Doing this (accessing the underlying socket via `s`) is definitely not the right way to do things.  If you want a quick workaround then try:
```py
self.a_writer.out_buf += premsg
```

Carglglz commented:
> Are you able to provide some code that shows how to reproduce this particular failure?

Yes here I checked out a branch from this one with the code at https://github.com/Carglglz/micropython/tree/asyncio-tls-tests/tests/net_inet/microdot_asyncio

In that dir in `hello_async_tls.py` to enable client authentication uncomment
```py
# ctx.verify_mode = _ssl.CERT_REQUIRED
ctx.load_cert_chain(cert, keyfile=key)
# ctx.load_verify_locations(cadata=cert)
```
run 
```
$ micropython hello_async_tls.py
Starting async server on 0.0.0.0:5000...
```
Then with `ab` or `curl`
```
$ab -n 1 -c 1 -v 3 https://127.0.0.1:5000/
This is ApacheBench, Version 2.3 <$Revision: 1901567 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 127.0.0.1 (be patient)...INFO: GET header ==
---
GET / HTTP/1.0
Host: 127.0.0.1:5000
User-Agent: ApacheBench/2.3
Accept: */*


---
SSL/TLS Handshake [Start] before/connect initialization
SSL/TLS State [connect] before/connect initialization
SSL/TLS State [connect] SSLv3 write client hello A
SSL/TLS State [connect] SSLv3 read server hello A
SSL/TLS State [connect] SSLv3 read server certificate A
SSL/TLS State [connect] SSLv3 read server certificate request A
SSL/TLS State [connect] SSLv3 read server done A
SSL/TLS State [connect] SSLv3 write client certificate A
SSL/TLS State [connect] SSLv3 write client key exchange A
SSL/TLS State [connect] SSLv3 write change cipher spec A
SSL/TLS State [connect] SSLv3 write finished A
SSL/TLS State [connect] SSLv3 flush data
SSL/TLS Alert [read] fatal:handshake failure
SSL handshake failed (1).
4688207532:error:1401E410:SSL routines:CONNECT_CR_FINISHED:sslv3 alert handshake failure:/System/Volumes/Data/SWE/macOS/BuildRoots/37599d3d49/Library/Caches/com.apple.xbs/Sources/libressl/libressl-56.60.4/libressl-2.8/ssl/ssl_pkt.c:1200:SSL alert number 40
..done


Server Software:
Server Hostname:        127.0.0.1
Server Port:            5000

Document Path:          /
Document Length:        0 bytes

Concurrency Level:      1
Time taken for tests:   0.713 seconds
Complete requests:      1
Failed requests:        0
Total transferred:      0 bytes
HTML transferred:       0 bytes
Requests per second:    1.40 [#/sec] (mean)
Time per request:       713.186 [ms] (mean)
Time per request:       713.186 [ms] (mean, across all concurrent requests)
Transfer rate:          0.00 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       0
Processing:   713  713   0.0    713     713
Waiting:        0    0   0.0      0       0
Total:        713  713   0.0    713     713
```

```
$ curl -k https://localhost:5000/
curl: (35) error:1401E410:SSL routines:CONNECT_CR_FINISHED:sslv3 alert handshake failure
```

Without `mbedtls_ssl_send_alert_message()` they will hang indefinitely until timeout or server shutdown...

> `self.a_writer.out_buf += premsg`

I've just tested this and it doesn't work. 

> Doing this (accessing the underlying socket via s) is definitely not the right way to do things

I'm aware, but I've not found any other way around (also it's only used once, just right after connection) 
Any hint on what exactly the second arg does? 👀

dpgeorge commented:
> > `self.a_writer.out_buf += premsg`
> 
> I've just tested this and it doesn't work.

Ah, sorry, I forgot about the second argument.  As you found out it's the maximum number of bytes to write.

peterzuger commented on extmod/modssl_mbedtls.c:
```diff
@@ -138,6 +156,41 @@ STATIC NORETURN void mbedtls_raise_error(int err) {
+        // The length of the string written (not including the terminated nul byte),
+        // or a negative err code.
+        if (ret > 0) {
+            mp_raise_ValueError(MP_ERROR_TEXT(xcbuf));
```
This needs to be changed to something like:

`mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("%s"), xcbuf);`

Otherwise this PR works great!

Carglglz replied:
Yes, I did notice that the error message wasn't properly handled and I guess garbage collected? But I didn't know how to fix it and I haven't looked into it yet, so thanks @peterzuger for testing this PR and providing the fix 🥇 , I will add it asap 👍🏼

dpgeorge commented:
@Carglglz sorry this hasn't received much attention lately.  The blocker on this is how to get compatibility with CPython.  If a module is named by the same name as a corresponding CPython module (in this case `ssl`) then we need to make sure it has the same functionality/usage, at least for the subset of the module that is implemented.  If we can't do that then we need to invent a new MicroPython-specific module for the functionality that is different to CPython.

Eg recently we added a new `deflate` module with MicroPython-specific functionality, and then implemented the CPython-compatible `zlib` and `gzip` modules on top of that.

For `ssl` the thing that stands out as different to CPython is loading certificate chains from a bytes object rather than a file, in `ssl.load_cert_chain()`.  It is very useful to be able to load them from a bytes/str object, and there have even been many requests for this feature in CPython.  See https://github.com/python/cpython/issues/60691 and even a fully working PR for that feature at https://github.com/python/cpython/pull/2449.  Unfortunately for us the outcome for CPython was:

> Unfortunately, closing because of four-year stalemate. The feature is useful by itself but found no support of the SSH part maintainer both from architectural standpoint ... and verification area creep

(See the end of the linked PR above.)

So... that leaves us in the situation where we need to do one of the following:
1. Break compatibility with CPython for the `ssl.load_cert_chain()` function, currently what this PR does.
2. Be compatible with CPython and support loading from a file on disk.
3. Invent a new module for SSL which is MicroPython-specific.

As mentioned above we want to maintain CPython compatibility for modules that are named after their CPython counterpart (one benefit of this is that tests will run under both CPython and MicroPython, making it much easier to validate behaviour of MicroPython).  So option (1) is out.

It seems a shame to have to go for option (3), invent a whole new module just so we can change one parameter of one function (ie allow `load_cert_chain()` to take a bytes/str instead of a file name).  On the other hand, if we have a new module then we can add other things like pre-allocation of SSL buffers, DTLS support, PSK support, etc.  So probably in the long term option (3) will need to happen regardless.

But in the short term, to get this PR over the line and merged, I suggest we go for option (2).  That means `load_cert_chain()` needs to be changed so that the `certfile` argument is a file name and the function loads the certificate from disk, using the `mp_vfs_open()` function (and `mp_get_stream`, `mp_stream_rw`).

@Carglglz do you want to attempt to make that change?

Carglglz commented:
@dpgeorge 

> 1. Break compatibility with CPython for the ssl.load_cert_chain() function, currently what this PR does.
>2. Be compatible with CPython and support loading from a file on disk.
>3. Invent a new module for SSL which is MicroPython-specific.

I agree with 2 and 3, but I also think MicroPython can support both (see latest commit https://github.com/micropython/micropython/commit/c1cafa2613716acf389cbf0b8aba0ff0d7fb16b5)

> using the mp_vfs_open() function (and mp_get_stream, mp_stream_rw).
>
> @Carglglz do you want to attempt to make that change?

Thanks for the pointers, I did try and it wasn't easy but luckily I found the `stream_readall` function in `py/stream.c` which I've reused in `read_file` function, but I think it would be better to expose `stream_readall`  in `py/stream.h` to avoid duplication? 
So please review the latest commit and let me know if there is anything I can do to improve it 👍🏼.

Also I've added a test in `sslcontext_server_client_files.py.` but there is something wrong with the unix-32-bit systems 
and `mbedtls + streams` e.g. this is the comment that I added in the tests that fail
```
# Asyncio + TLS in 32-bit systems fail.
# Reason: UNKNOWN
# Possible cause:
#   Incompatible ciphersuites + streams.
#   i.e. error OSError: (-29568, 'MBEDTLS_ERR_SSL_NO_CIPHER_CHOSEN')
# Action: SKIP for now.
if sys.maxsize == 2147483647:
    print("SKIP")
    raise SystemExit
```
But I have no idea why it fails or how to debug it... maybe I'm missing a `MICROPY` macro? It would be nice to get a diff between builds... 🤔

dpgeorge commented:
> Thanks for the pointers, I did try and it wasn't easy but luckily I found the `stream_readall` function in `py/stream.c` which I've reused in `read_file` function, but I think it would be better to expose `stream_readall` in `py/stream.h` to avoid duplication?
> So please review the latest commit and let me know if there is anything I can do to improve it 👍🏼.

Upon review of this I think a much simpler way to do it is just call `.read()` on the vfs file.  Eg:
```c
// file = open(args[0], "rb")
mp_obj_t fk_args[2] = {
    args[0].u_obj,
    MP_OBJ_NEW_QSTR(MP_QSTR_rb),
};
mp_obj_t keyf = mp_vfs_open(2, &fk_args[0], (mp_map_t *)&mp_const_empty_map);

// data = file.read()
mp_obj_t dest[2];
mp_load_method(keyf, MP_QSTR_read, dest);
mp_obj_t data = mp_call_method_n_kw(0, 0, dest);

// file.close()
mp_load_method(keyf, MP_QSTR_close, dest);
mp_call_method_n_kw(0, 0, dest);
```

I would then wrap that in a function and use it in the 3 locations where it's needed.

Carglglz commented:
> Upon review of this I think a much simpler way to do it is just call .read() on the vfs file. Eg:

I've just tested this and it works as expected! Thanks! 👍🏼 
I will update the latest commit asap.

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -264,6 +409,73 @@ STATIC void ssl_context_load_key(mp_obj_ssl_context_t *self, mp_obj_t key_obj, m
+
+    if (pos_args[1] != mp_const_none) {
+        mp_check_self(mp_obj_is_str_or_bytes(pos_args[1])); // CERT
+        mp_check_self(mp_obj_is_str_or_bytes(args[0].u_obj)); // KEY
```
These `mp_check_self()` calls are not needed.  The type will be checked automatically when the str/bytes value is extracted.

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -276,6 +454,41 @@ STATIC void ssl_context_load_cadata(mp_obj_ssl_context_t *self, mp_obj_t cadata_
+    if (args[0].u_obj != mp_const_none) {
+        mp_obj_t cadata;
+
+        if (!(mp_obj_is_type(args[0].u_obj, &mp_type_bytes))) {
```
No need to test for a bytes object, just call `read_file` unconditionally, eg:
```c
ssl_context_load_cadata(self, read_file(args[0].u_obj));
```

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -276,6 +454,41 @@ STATIC void ssl_context_load_cadata(mp_obj_ssl_context_t *self, mp_obj_t cadata_
+
+    }
+    // cadata
+    else if (args[1].u_obj != mp_const_none) {
```
I think CPython allows specifying both cafile and cadata at the same time, so this can just be `if` (not `else if`).

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -369,6 +594,18 @@ STATIC mp_obj_t ssl_socket_make_new(mp_obj_ssl_context_t *ssl_context, mp_obj_t
+    #ifdef MICROPY_SSL_MBEDTLS_EXTRAS
+    uint32_t flags = 0;
+    int *cid = &ssl_context->cipherid[0];
+    /* printf("cipherid: %d\n", *cid); */
```
Please remove debugging prints like this that are commented out.

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -454,6 +723,36 @@ STATIC mp_uint_t socket_read(mp_obj_t o_in, void *buf, mp_uint_t size, int *errc
         o->last_error = ret;
     }
+
+    if (
```
This error handling code looks exactly the same as that in `socket_write` below.  So please factor them out into a helper function to reduce code duplication.

dpgeorge commented on ports/esp32/esp32_common.cmake:
```diff
@@ -63,6 +63,8 @@ list(APPEND MICROPY_SOURCE_PORT
     fatfs_port.c
     help.c
+	modtime.c
+	mbedtls/mbedtls_port.c
```
Please use spaces instead of tabs here.

Carglglz commented:
@dpgeorge Thanks for reviewing this. I think it's now fixed, but let me know if you see something else 👍🏼 

I've updated https://github.com/micropython/micropython/pull/12432 accordingly. 

Also there are a few PR in `micropython-lib` to support this, that may be nice to have .e.g:

- [ ] https://github.com/micropython/micropython-lib/pull/722
- [ ] https://github.com/micropython/micropython-lib/pull/724 & https://github.com/micropython/micropython-lib/pull/752

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -114,6 +122,18 @@ def open_connection(host, port):
+        if not _ssl:
+            raise ValueError("SSL not supported")
+        if ssl is True:
+            ssl = _ssl.SSLContext(_ssl.PROTOCOL_TLS_CLIENT)
```
Instead of importing ssl at the top of this file, it can be delayed to this point.  Then this function can be used with a custom SSLContext passed in as the `ssl` argument and the port doesn't even need to have the `ssl` module.

I suggest:
```py
if ssl is True:
    import ssl as _ssl

    ssl = _ssl.SSLContext(_ssl.PROTOCOL_TLS_CLIENT)
```

If the `ssl` module doesn't exist then it'll raise an import error, which I think is good enough to inform the user that SSL is not supported because the `ssl` module does not exist.

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -138,6 +180,70 @@ STATIC NORETURN void mbedtls_raise_error(int err) {
+        // handshake failure notify peer
+        // with proper error code and raise mp error with mbedtls_raise_error
+        #ifdef MICROPY_SSL_MBEDTLS_EXTRAS
+        ssl_handle_handshake_error(sslsock, errcode);
```
It looks like this function can be merged here directly, because it's only called from this location.

Then, common code can be factored out with the #else block below.  Eg the call to `mbedtls_ssl_send_alert_message()`, and maybe the `mbedtls_ssl_free()` and the final `mbedtls_raise_error()`.

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -138,6 +180,70 @@ STATIC NORETURN void mbedtls_raise_error(int err) {
+        #if MBEDTLS_VERSION_NUMBER >= 0x03000000
+        (*errcode < 0) & (mbedtls_ssl_is_handshake_over(&sslsock->ssl) == 0) & (*errcode != MBEDTLS_ERR_SSL_CONN_EOF)
+        #else
+        (*errcode < 0) & (*errcode != MBEDTLS_ERR_SSL_CONN_EOF)
```
I think these `&` need to be `&&`?

Carglglz commented:
@dpgeorge Thanks for reviewing again! I think I've addressed all the comments now. 👍🏼

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -100,7 +103,7 @@ def drain(self):
 #
 # async
-def open_connection(host, port):
+async def open_connection(host, port, ssl=None, server_hostname=None):
```
This `async` keyword should not be here, because this function only calls `yield`.

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -114,6 +117,18 @@ def open_connection(host, port):
+        s = ssl.wrap_socket(s, server_hostname=server_hostname, do_handshake_on_connect=False)
+        s.setblocking(False)
+        ss = Stream(s)
+        yield core._io_queue.queue_write(s)
```
These two lines here can probably be de-dented, and then the `ss = Stream(s)` and `yield core._io_queue.queue_write(s)` lines above (at line 113 and 119) be removed.  Then this bit here looks like:
```py
    ...
    s.setblocking(False)
ss = Stream(s)
yield core._io_queue.queue_write(s)
return ss, ss
```
(The call to wrap_socket doesn't need the socket to be connected when `do_handshake_on_connect=False`.)

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -156,14 +171,23 @@ async def _serve(self, s, cb):
+                    s2 = ssl.wrap_socket(s2, server_side=True, do_handshake_on_connect=False)
+                    s2.setblocking(False)
+                except OSError as e:
+                    sys.print_exception(e)
```
Instead of importing sys, use `core.sys.print_exception(e)`.  It will be slightly smaller code and use slightly less RAM (in the globals dict of this module).

dpgeorge replied:
Also you can probably move the `s2.setblocking(False)` to after this `if ssl:` block, and remove line 175 above.

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -264,6 +425,50 @@ STATIC void ssl_context_load_key(mp_obj_ssl_context_t *self, mp_obj_t key_obj, m
+        if (!(mp_obj_is_type(args[0].u_obj, &mp_type_bytes))) {
+
+            pkey = read_file(args[0].u_obj);
+
```
Please remove all the empty lines here.  Try to follow the style of existing code.

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -300,6 +540,12 @@ STATIC const mp_rom_map_elem_t ssl_context_locals_dict_table[] = {
     #if MICROPY_PY_SSL_FINALISER
     { MP_ROM_QSTR(MP_QSTR___del__), MP_ROM_PTR(&ssl_context___del___obj) },
     #endif
+    #ifdef MICROPY_SSL_MBEDTLS_EXTRAS
```
I don't think we need this config option.  We want to migrate all users of this ssl module to use these new methods to load certificates, so they should be unconditionally available.

Carglglz replied:
I added this because not every board (i.e. in stm32 port) could fit the extra bytes and the build process failed...
So this seemed the best option, to be able to enable these features at port or board level...

Carglglz replied:
I've removed the macro and it seems the build process is OK in stm32 port now 👍🏼

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -300,6 +540,12 @@ STATIC const mp_rom_map_elem_t ssl_context_locals_dict_table[] = {
     { MP_ROM_QSTR(MP_QSTR___del__), MP_ROM_PTR(&ssl_context___del___obj) },
     #endif
+    #ifdef MICROPY_SSL_MBEDTLS_EXTRAS
+    { MP_ROM_QSTR(MP_QSTR_get_ciphers), MP_ROM_PTR(&mod_ssl_get_ciphers_obj)},
```
The function identifiers should match the Python name.  Eg here it should be:
```c
{ MP_ROM_QSTR(MP_QSTR_get_ciphers), MP_ROM_PTR(&ssl_context_get_ciphers_obj)},
```

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -567,6 +859,9 @@ STATIC const mp_rom_map_elem_t ssl_socket_locals_dict_table[] = {
     #endif
     { MP_ROM_QSTR(MP_QSTR_getpeercert), MP_ROM_PTR(&mod_ssl_getpeercert_obj) },
+    #ifdef MICROPY_SSL_MBEDTLS_EXTRAS
+    { MP_ROM_QSTR(MP_QSTR_cipher), MP_ROM_PTR(&mod_ssl_cipher_obj) },
```
What is this method needed for?

Carglglz replied:
To return the current TLS protocol and ciphersuite being used see, https://docs.python.org/3/library/ssl.html#ssl.SSLSocket.cipher, 
CPython returns a 3-element tuple, but in `mbedtls` there is no function for the third element .i.e 
> the number of secret bits being used

But I think it's useful enough even if it doesn't match CPython and I don't think it would break any compatibility

Carglglz commented:
@dpgeorge thanks for the tips again! I think the comments are now addressed, but let me know if I'm missing anything else 👍🏼

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -138,6 +171,51 @@ STATIC NORETURN void mbedtls_raise_error(int err) {
+            // The certificate may have been rejected for several reasons.
+            flags = mbedtls_ssl_get_verify_result(&sslsock->ssl);
+
+        } else if (*errcode == MBEDTLS_ERR_SSL_NO_CLIENT_CERTIFICATE) {
```
It would make more sense to put this if-block first, then the `MBEDTLS_ERR_X509_CERT_VERIFY_FAILED` block second, and then the latter can be merged with the addition if-block below.

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -138,6 +171,51 @@ STATIC NORETURN void mbedtls_raise_error(int err) {
+        }
+
+        if (*errcode == MBEDTLS_ERR_X509_CERT_VERIFY_FAILED) {
+            char xcbuf[512];
```
Does this need to be 512 bytes?  That's quite a lot of stack space.  Can it be 64 without losing information?  How about 128?

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -167,6 +246,9 @@ STATIC mp_obj_t ssl_context_make_new(const mp_obj_type_t *type_in, size_t n_args
     mbedtls_debug_set_threshold(3);
     #endif
+    #ifdef MICROPY_MBEDTLS_PLATFORM_TIME_ALT
+    mbedtls_platform_set_time(platform_mbedtls_time);
```
I suggest putting this function call in `esp32/main.c`, because it's very specific to the esp32 implementation.  All other ports use `MBEDTLS_PLATFORM_TIME_MACRO`, which is a simpler way to configure the time function.

(And if/when esp32 switches to 1970 as the Epoch, we can just delete this bit.)

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -236,6 +318,55 @@ STATIC mp_obj_t ssl_context___del__(mp_obj_t self_in) {
+STATIC mp_obj_t ssl_context_get_ciphers(mp_obj_t self_in) {
+    mp_obj_t list = mp_obj_new_list(0, NULL);
+    const int *cipher_list = mbedtls_ssl_list_ciphersuites();
+    while (*cipher_list) {
```
This would be better as a for loop:
```c
for (const int *cipher_list = mbedtls_ssl_list_ciphersuites(); *cipher_list; ++cipher_list) {
    ...
}
```

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -236,6 +318,55 @@ STATIC mp_obj_t ssl_context___del__(mp_obj_t self_in) {
+
+STATIC mp_obj_t ssl_context_set_ciphers(mp_obj_t self_in, mp_obj_t ciphersuite) {
+    // check that ciphersuite is a list
+    if (mp_obj_is_type(ciphersuite, &mp_type_list)) {
```
Better to use `mp_obj_get_array(ciphersuite, &len, &items)`.  That will check for tuple or list and raise an exception if not.

Carglglz replied:
`mp_obj_get_array(ciphersuite, &len, &items)`

Not sure how to implement this, i.e. do I need to declare `len` and `items` first? or pass `ciphersuite->len` etc?

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -264,6 +395,40 @@ STATIC void ssl_context_load_key(mp_obj_ssl_context_t *self, mp_obj_t key_obj, m
 }
 
+
+STATIC mp_obj_t ssl_context_load_certchain(size_t n_args, const mp_obj_t *pos_args,
```
CPython's signature of this method is:
```py
SSLContext.load_cert_chain(certfile, keyfile=None, password=None)
```
But it looks like mbedtls requires both cert and key.  So MicroPython's signature should be:
```py
SSLContext.load_cert_chain(certfile, keyfile)
```
Ie both arguments are required.

So please update this code here to implement this signature:
```c
STATIC mp_obj_t ssl_context_load_cert_chain(mp_obj_t self_in, mp_obj_t certfile, mp_obj_t keyfile) {
    ...
}
STATIC MP_DEFINE_CONST_FUN_OBJ_3(ssl_context_load_cert_chain_obj, ssl_context_load_cert_chain);
```

Carglglz replied:
I think this breaks CPython compatibility? 

```C
STATIC mp_obj_t ssl_context_load_cert_chain(mp_obj_t self_in, mp_obj_t certfile, mp_obj_t keyfile) {
    ...
}
STATIC MP_DEFINE_CONST_FUN_OBJ_3(ssl_context_load_cert_chain_obj, ssl_context_load_cert_chain);
```
This doesn't allow keyword args, e.g. this in MicroPython will fail
```py
SSLContext.load_cert_chain("cert.der", keyfile="key.der")
```

I modified `keyfile` arg and made it `MP_ARG_REQUIRED`, this should allow both forms, right?

```py
SSLContext.load_cert_chain(certfile, keyfile=None)
```
or
```py
SSLContext.load_cert_chain(certfile, keyfile)
 ```

dpgeorge commented on ports/esp32/mbedtls/mbedtls_config.h:
```diff
@@ -0,0 +1,4 @@
+// Time mbedtls_platform
+#define MBEDTLS_HAVE_ASM
+
+time_t platform_mbedtls_time(time_t *timer);
```
I don't think this file will be picked up by the build, the IDF has its own config file.  So just remove it.

dpgeorge commented on ports/esp32/mbedtls/mbedtls_port.c:
```diff
@@ -0,0 +1,24 @@
+#include "shared/timeutils/timeutils.h"
+
+
+#ifdef MICROPY_MBEDTLS_PLATFORM_TIME_ALT
```
You can use `#if CONFIG_MBEDTLS_PLATFORM_TIME_ALT` instead.

Carglglz committed:
extmod/asyncio: Add ssl support with SSLContext.

This adds asyncio ssl support with SSLContext and the corresponding
tests in `tests/net_inet` and `tests/multi_net`.

Note that not doing the handshake on connect will delegate the handshake to
the following `mbedtls_ssl_read/write` calls.  However if the handshake
fails when a client certificate is required and not presented by the peer,
it needs to be notified of this handshake error (otherwise it will hang
until timeout if any).  Finally at MicroPython side raise the proper
mbedtls error code and message.

Signed-off-by: Carlos Gil <carlosgilglez@gmail.com>

Carglglz commented on extmod/modssl_mbedtls.c:
```diff
@@ -59,6 +70,7 @@ typedef struct _mp_obj_ssl_context_t {
     mbedtls_x509_crt cert;
     mbedtls_pk_context pkey;
     int authmode;
+    int cipherid [];
```
I've tried to make these changes but failed, latest changes are how far I got before hitting errors. I may be missing something obvious but I've no clue honestly. 😕

Carglglz replied:
OK, I've figured out what I was doing wrong, I think it's resolved now.

Carglglz commented:
@dpgeorge I've just realised this was getting a bit messy to review, maintain and update so I've reorganised and split the PR in:

1. `SSLContext` - [ ] https://github.com/micropython/micropython/pull/13098  --> Here we can further discuss: 
      - [ ] `SSLContext.set_ciphers` implementation (see comments above) 
      - [ ] Buffer size/implementation for cert verify errors.
      - [ ] `SSLContext.load_cert_chain(certfile, keyfile=)` MicroPython signature.

2. `asyncio-tls`  -[ ] https://github.com/micropython/micropython/pull/11897   --> This PR, and I think all the comments related to asyncio implementation are addressed now.  If you spot anything else, let me know.

---
3 & 4 `<port>_mbedtls_time`--> Cert time validation is unrelated to `SSLContext` or `asyncio-tls` so they can be reviewed and merged independently.

3. `unix_mbedtls_time` --> This just enables the necessary config options so it should be easy to review/merge.
    - [ ] https://github.com/micropython/micropython/pull/13099

4. `esp32_mbedtls_time` --> It's just a WIP, I don't know how to implement this outside `modssl_mbedtls.c`  as you suggested so feel free to make any changes...
   - [ ]https://github.com/micropython/micropython/pull/13100

dpgeorge commented on tests/multi_net/ssl_get_peer_cert.py:
```diff
@@ -0,0 +1,61 @@
+# $ openssl rsa -in rsa_key.pem -out rsa_key.der -outform DER
+# $ openssl x509 -in rsa_cert.pem -out rsa_cert.der -outform DER
+#
+# Then convert to hex format, eg using binascii.hexlify(data).
```
Instead of having this text in each file that uses the certificates, I suggest putting this in `tests/README.md` (in a new section in that file), and then put a single-line comment in the tests like this:

    # These are test certificates.  See tests/README.md for details.

dpgeorge commented:
@Carglglz can you please rebase this on the latest master?  Then it's hopefully ready to be merged.

Carglglz commented:
@dpgeorge I've update the tests with the `os.stat` logic too 👍🏼

dpgeorge commented on tests/multi_net/asyncio_tcp_tls_server_client.py:
```diff
@@ -0,0 +1,132 @@
+    await tcp_client(b"client data")
+    multitest.next()
+    print("TLS CLIENT")
+    await tls_client(b"client data")
```
This is doing a non-SSL connection followed by an SSL-connection.  But what exactly is that testing?  This feels more like an integration test rather than a unit test.  Did you see issues with non-SSL followed by SSL connections?

Also there's the corresponding `asyncio_tls_tcp_server_client.py` test which flips the order of these connections.  Is it necessary to have both of theses sets of tests?

If it is necessary to have this kind of test I'd suggest just a single test (not two separate files) that does non-SSL then SSL then non-SSL again.

Carglglz replied:
>  But what exactly is that testing? This feels more like an integration test rather than a unit test. Did you see issues with non-SSL followed by SSL connections?

That was prior to https://github.com/micropython/micropython/pull/12224 see https://github.com/micropython/micropython/pull/11897#discussion_r1290401298

But I guess now that it's fixed, they are not necessary any more?

dpgeorge replied:
Oh, I see.

Now that bug is fixed, testing combinations of non-SSL and SSL is no longer needed.

So the tests here can just be for SSL.

dpgeorge commented on tests/multi_net/asyncio_tls_server_client.py:
```diff
@@ -0,0 +1,76 @@
+
+async def handle_connection(reader, writer):
+    # Test that peername exists (but don't check its value, it changes)
+    writer.get_extra_info("peername")
```
I don't think this line is needed, the `get_extra_info()` method is tested already in an existing test.

dpgeorge commented on tests/net_inet/asyncio_http_https_opc.py:
```diff
@@ -0,0 +1,81 @@
+
+client_ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+client_ctx.verify_mode = ssl.CERT_REQUIRED
+client_ctx.load_verify_locations(cafile=ca_cert_chain)
```
This test (and the other 3 added to the `net_inet` directory) don't run on CPython (at least not 3.11.6) because CPython's `load_verify_locations()` requires the file to be in PEM format, not DER.

The fix for that is to load the data from the DER here and then pass it in as `cadata`.

But, that still doesn't seem to have the correct behaviour because I would think that you don't need the line below (the call to `load_default_certs()`) because it should be enough to load just the DER certificate from `mpycert.der` in order to access micropython.org.  But if I remove the `load_default_certs()` then CPython cannot connect to micropython.org.  If I leave `load_default_certs()` in place then I can just remove this `load_verify_locations()` call and CPython works.

So it seems this test is not working as expected on CPython.

Carglglz replied:
> The fix for that is to load the data from the DER here and then pass it in as cadata.

I agree

> But, that still doesn't seem to have the correct behaviour because I would think that you don't need the line below (the call to load_default_certs()) because it should be enough to load just the DER certificate from mpycert.der in order to access micropython.org. But if I remove the load_default_certs() then CPython cannot connect to micropython.org. If I leave load_default_certs() in place then I can just remove this load_verify_locations() call and CPython works.

I could reproduce this, and it seems `mbedtls` is a bit more "permissive" than "OpenSSL" when it comes to certificate validation, i.e. `mbedtls` is OK using "intermediate" CA certificates, while "OpenSSL" wants the "ultimate" root CA certificate.

The fix is using this root CA certificate that both MicroPython and CPython accept 👍🏼 

FWIW I got this root CA certificate from https://github.com/micropython/micropython-lib/pull/633 so if you have time you may want to have a look.

dpgeorge replied:
OK, using the ISRG certificate is a good idea.  Except it doesn't seem to work for me, running that test I get:


CPython:
```
ssl.SSLError: not enough data: cadata does not contain a certificate (_ssl.c:4020)
```

MicroPython:
```
ValueError: invalid cert
```

dpgeorge commented on tests/net_inet/asyncio_http_https_opc.py:
```diff
@@ -0,0 +1,81 @@
+
+async def http_https_get(url1, url2, port, sslctx):
+    await http_get(url1)
+    await https_get(url2, port, client_ctx)
```
Is it necessary to test non-SSL followed by SSL?  Non-SSL connections are already tested by existing tests, so it'd be good if this test just did the SSL test, ie just called `https_get()`.

Then the `asyncio_https_http_opc.py` test is not needed.

dpgeorge commented on tests/net_inet/asyncio_tls_open_connection.py:
```diff
@@ -0,0 +1,51 @@
+# Test simple HTTPS request with uasyncio.open_connection()
```
I don't think this test is needed.  It should be enough to have `asyncio_tls_open_connection_readline.py`.

dpgeorge commented on tests/multi_net/asyncio_tcp_tls_server_client.py.exp:
```diff
@@ -0,0 +1,20 @@
+--- instance0 ---
```
this file is no longer needed

dpgeorge commented on tests/multi_net/asyncio_tls_tcp_server_client.py.exp:
```diff
@@ -0,0 +1,20 @@
+--- instance0 ---
```
this file is no longer needed

dpgeorge commented:
Merged!  Thanks again @Carglglz for sticking with this and getting it through.

Carglglz commented:
@dpgeorge Thank you for the support and for helping to get this through! 👍🏼
